File: go/template_parts.go
package vbml

import (
	"errors"
	"strings"
)

const (
	errMutuallyExclusiveTemplateSources = "component template and templateParts are mutually exclusive"
	templateSpaceCell                   = "{0}"
)

type inlineSegment struct {
	Atomic bool
	Cells  []string
}

type templateTokenKind int

const (
	templateTokenWord templateTokenKind = iota
	templateTokenSpace
	templateTokenNewline
)

type templateToken struct {
	kind     templateTokenKind
	segments []inlineSegment
}

type templateLineState struct {
	line   string
	length int
}

func resolveComponentTemplateLines(width int, props map[string]any, component Component) ([]string, error) {
	parts, err := normalizeComponentTemplate(component)
	if err != nil {
		return nil, err
	}

	return resolveTemplateLines(width, props, parts), nil
}

func resolveTemplateLines(width int, props map[string]any, parts []TemplatePart) []string {
	for index, part := range parts {
		parts[index] = preprocessTemplatePart(props, part)
	}

	return wrapTemplateTokens(width, buildTemplateTokens(parts))
}

func normalizeComponentTemplate(component Component) ([]TemplatePart, error) {
	if component.Template != "" && len(component.TemplateParts) > 0 {
		return nil, errors.New(errMutuallyExclusiveTemplateSources)
	}

	if len(component.TemplateParts) > 0 {
		return normalizeTemplateParts(component.TemplateParts), nil
	}

	return []TemplatePart{{
		Template: component.Template,
		Wrap:     TemplateWrapNormal,
	}}, nil
}

func normalizeTemplateParts(parts []TemplatePart) []TemplatePart {
	if len(parts) == 0 {
		return nil
	}

	normalized := make([]TemplatePart, 0, len(parts))
	for _, part := range parts {
		normalized = append(normalized, TemplatePart{
			Template: part.Template,
			Wrap:     normalizeTemplateWrap(part.Wrap),
		})
	}
	return normalized
}

func normalizeTemplateWrap(wrap TemplateWrap) TemplateWrap {
	if wrap == TemplateWrapNever {
		return TemplateWrapNever
	}

	return TemplateWrapNormal
}

func preprocessTemplatePart(props map[string]any, part TemplatePart) TemplatePart {
	return TemplatePart{
		Template: SanitizeSpecialCharacters(
			parseProps(props, emojisToCharacterCodes(part.Template)),
		),
		Wrap: normalizeTemplateWrap(part.Wrap),
	}
}

func tokenizeDisplayCells(template string) []string {
	runes := []rune(template)
	cells := make([]string, 0, len(runes))

	for index := 0; index < len(runes); index++ {
		current := string(runes[index])
		if current == "{" {
			closeIndex := index + 1
			for closeIndex < len(runes) && string(runes[closeIndex]) != "}" {
				closeIndex++
			}
			if closeIndex < len(runes) {
				cells = append(cells, string(runes[index:closeIndex+1]))
				index = closeIndex
				continue
			}
		}

		if current == " " {
			cells = append(cells, templateSpaceCell)
			continue
		}

		cells = append(cells, current)
	}

	return cells
}

func buildTemplateTokens(parts []TemplatePart) []templateToken {
	var tokens []templateToken
	var currentSegments []inlineSegment

	pushWordToken := func() {
		if len(currentSegments) == 0 {
			return
		}
		tokens = append(tokens, createWordToken(currentSegments))
		currentSegments = nil
	}

	pushSegment := func(atomic bool, cells []string) {
		if len(cells) == 0 {
			return
		}
		currentSegments = append(currentSegments, inlineSegment{
			Atomic: atomic,
			Cells:  cloneCells(cells),
		})
	}

	appendNormalCells := func(cells []string) {
		var currentCells []string

		flushCurrentCells := func() {
			pushSegment(false, currentCells)
			currentCells = nil
		}

		for _, cell := range cells {
			switch cell {
			case "\n":
				flushCurrentCells()
				pushWordToken()
				tokens = append(tokens, templateToken{kind: templateTokenNewline})
			case templateSpaceCell:
				flushCurrentCells()
				pushWordToken()
				tokens = append(tokens, templateToken{kind: templateTokenSpace})
			default:
				currentCells = append(currentCells, cell)
			}
		}

		flushCurrentCells()
	}

	appendAtomicCells := func(cells []string) {
		var currentCells []string

		flushCurrentCells := func() {
			pushSegment(true, currentCells)
			currentCells = nil
		}

		for _, cell := range cells {
			if cell == "\n" {
				flushCurrentCells()
				pushWordToken()
				tokens = append(tokens, templateToken{kind: templateTokenNewline})
				continue
			}

			currentCells = append(currentCells, cell)
		}

		flushCurrentCells()
	}

	for _, part := range parts {
		cells := tokenizeDisplayCells(part.Template)
		if normalizeTemplateWrap(part.Wrap) == TemplateWrapNever {
			appendAtomicCells(cells)
			continue
		}
		appendNormalCells(cells)
	}

	pushWordToken()

	return tokens
}

func wrapTemplateTokens(width int, tokens []templateToken) []string {
	lines := []templateLineState{{
		line:   "",
		length: 0,
	}}
	chunkedTokens := chunkWordTokens(width, tokens)

	for index, token := range chunkedTokens {
		lastIndex := len(lines) - 1
		lineLength := lines[lastIndex].length
		emptyLine := lines[lastIndex].line == ""

		if token.kind == templateTokenNewline && index > 0 && chunkedTokens[index-1].kind == templateTokenNewline {
			lines = append(lines, templateLineState{})
			continue
		}

		if token.kind == templateTokenNewline {
			lines[lastIndex] = templateLineState{
				line:   lines[lastIndex].line,
				length: width,
			}
			continue
		}

		if token.kind == templateTokenSpace {
			if lineLength+1 > width {
				continue
			}

			lines[lastIndex] = templateLineState{
				line:   lines[lastIndex].line + templateSpaceCell,
				length: lineLength + 1,
			}
			continue
		}

		wordWidth := measureWordWidth(token)
		if width >= wordWidth+lineLength && !emptyLine {
			lines[lastIndex] = templateLineState{
				line:   lines[lastIndex].line + wordTokenToString(token),
				length: lineLength + wordWidth,
			}
			continue
		}

		lines = append(lines, templateLineState{
			line:   wordTokenToString(token),
			length: wordWidth,
		})
	}

	startIndex := 0
	if lines[0].line == "" {
		startIndex = 1
	}

	result := make([]string, 0, len(lines)-startIndex)
	for _, line := range lines[startIndex:] {
		result = append(result, line.line)
	}
	return result
}

func chunkWordTokens(width int, tokens []templateToken) []templateToken {
	var chunkedTokens []templateToken

	for _, token := range tokens {
		if token.kind != templateTokenWord || measureWordWidth(token) <= width {
			chunkedTokens = append(chunkedTokens, token)
			continue
		}

		remaining := &token
		for remaining != nil && measureWordWidth(*remaining) > width {
			prefix, remainder := takeWordPrefix(*remaining, width)
			if prefix == nil {
				break
			}

			chunkedTokens = append(chunkedTokens, *prefix)
			remaining = remainder
		}

		if remaining != nil {
			chunkedTokens = append(chunkedTokens, *remaining)
		}
	}

	return chunkedTokens
}

func takeWordPrefix(token templateToken, maxWidth int) (*templateToken, *templateToken) {
	if maxWidth <= 0 {
		remainder := createWordToken(cloneSegments(token.segments))
		return nil, &remainder
	}

	var prefixSegments []inlineSegment
	var remainderSegments []inlineSegment
	consumedWidth := 0

	appendRemainder := func(startIndex int) {
		remainderSegments = append(remainderSegments, cloneSegments(token.segments[startIndex:])...)
	}

	for index, segment := range token.segments {
		segmentWidth := len(segment.Cells)
		if consumedWidth+segmentWidth <= maxWidth {
			prefixSegments = append(prefixSegments, cloneSegment(segment))
			consumedWidth += segmentWidth
			continue
		}

		if segment.Atomic {
			if consumedWidth == 0 {
				prefixSegments = append(prefixSegments, inlineSegment{
					Atomic: true,
					Cells:  cloneCells(segment.Cells[:maxWidth]),
				})

				if segmentWidth > maxWidth {
					remainderSegments = append(remainderSegments, inlineSegment{
						Atomic: true,
						Cells:  cloneCells(segment.Cells[maxWidth:]),
					})
				} else {
					remainderSegments = append(remainderSegments, cloneSegment(segment))
				}

				appendRemainder(index + 1)
			} else {
				remainderSegments = append(remainderSegments, cloneSegment(segment))
				appendRemainder(index + 1)
			}

			return newWordPrefix(prefixSegments), newWordPrefix(remainderSegments)
		}

		remainingWidth := maxWidth - consumedWidth
		if remainingWidth > 0 {
			prefixSegments = append(prefixSegments, inlineSegment{
				Atomic: false,
				Cells:  cloneCells(segment.Cells[:remainingWidth]),
			})
			remainderSegments = append(remainderSegments, inlineSegment{
				Atomic: false,
				Cells:  cloneCells(segment.Cells[remainingWidth:]),
			})
		} else {
			remainderSegments = append(remainderSegments, cloneSegment(segment))
		}

		appendRemainder(index + 1)
		return newWordPrefix(prefixSegments), newWordPrefix(remainderSegments)
	}

	prefix := createWordToken(prefixSegments)
	return &prefix, nil
}

func newWordPrefix(segments []inlineSegment) *templateToken {
	if len(segments) == 0 {
		return nil
	}

	token := createWordToken(segments)
	return &token
}

func createWordToken(segments []inlineSegment) templateToken {
	return templateToken{
		kind:     templateTokenWord,
		segments: cloneSegments(segments),
	}
}

func cloneSegments(segments []inlineSegment) []inlineSegment {
	cloned := make([]inlineSegment, 0, len(segments))
	for _, segment := range segments {
		cloned = append(cloned, cloneSegment(segment))
	}
	return cloned
}

func cloneSegment(segment inlineSegment) inlineSegment {
	return inlineSegment{
		Atomic: segment.Atomic,
		Cells:  cloneCells(segment.Cells),
	}
}

func cloneCells(cells []string) []string {
	return append([]string(nil), cells...)
}

func measureWordWidth(token templateToken) int {
	width := 0
	for _, segment := range token.segments {
		width += len(segment.Cells)
	}
	return width
}

func wordTokenToString(token templateToken) string {
	var builder strings.Builder
	for _, segment := range token.segments {
		for _, cell := range segment.Cells {
			builder.WriteString(cell)
		}
	}
	return builder.String()
}

File: php/src/TemplateParts.php
<?php

namespace Vestaboard\Vbml;

/**
 * Helpers for normalized template-part wrapping.
 */
class TemplateParts
{
    private const SPACE_CELL = '{0}';

    public static function resolveTemplateLines(
        int $width,
        array $props,
        string $template
    ): array {
        $parts = array_map(
            fn(array $part) => self::preprocessTemplatePart($props, $part),
            self::normalizeLegacyTemplate($template)
        );

        return self::wrapTemplateTokens($width, self::buildTemplateTokens($parts));
    }

    private static function normalizeLegacyTemplate(string $template): array
    {
        return [[
            'template' => $template,
            'wrap' => 'normal',
        ]];
    }

    private static function preprocessTemplatePart(array $props, array $part): array
    {
        return [
            'template' => SanitizeSpecialCharacters::sanitize(
                ParseProps::parse($props, EmojisToCharacterCodes::convert($part['template']))
            ),
            'wrap' => $part['wrap'],
        ];
    }

    private static function tokenizeDisplayCells(string $template): array
    {
        $cells = [];
        $characters = mb_str_split($template);
        $count = count($characters);

        for ($index = 0; $index < $count; $index++) {
            $current = $characters[$index];

            if ($current === '{') {
                $closeIndex = $index + 1;
                while ($closeIndex < $count && $characters[$closeIndex] !== '}') {
                    $closeIndex++;
                }

                if ($closeIndex < $count) {
                    $cells[] = implode('', array_slice(
                        $characters,
                        $index,
                        $closeIndex - $index + 1
                    ));
                    $index = $closeIndex;
                    continue;
                }
            }

            if ($current === ' ') {
                $cells[] = self::SPACE_CELL;
                continue;
            }

            $cells[] = $current;
        }

        return $cells;
    }

    private static function cloneSegment(array $segment): array
    {
        return [
            'atomic' => $segment['atomic'],
            'cells' => array_values($segment['cells']),
        ];
    }

    private static function cloneSegments(array $segments): array
    {
        return array_map(fn(array $segment) => self::cloneSegment($segment), $segments);
    }

    private static function createWordToken(array $segments): array
    {
        return [
            'kind' => 'word',
            'segments' => $segments,
        ];
    }

    private static function buildTemplateTokens(array $parts): array
    {
        $tokens = [];
        $currentSegments = [];

        $pushWordToken = function () use (&$tokens, &$currentSegments): void {
            if (empty($currentSegments)) {
                return;
            }

            $tokens[] = self::createWordToken($currentSegments);
            $currentSegments = [];
        };

        $pushSegment = function (bool $atomic, array $cells) use (&$currentSegments): void {
            if (empty($cells)) {
                return;
            }

            $currentSegments[] = [
                'atomic' => $atomic,
                'cells' => array_values($cells),
            ];
        };

        $appendNormalCells = function (array $cells) use (&$tokens, $pushSegment, $pushWordToken): void {
            $currentCells = [];

            $flushCurrentCells = function () use (&$currentCells, $pushSegment): void {
                $pushSegment(false, $currentCells);
                $currentCells = [];
            };

            foreach ($cells as $cell) {
                if ($cell === "\n") {
                    $flushCurrentCells();
                    $pushWordToken();
                    $tokens[] = ['kind' => 'newline'];
                    continue;
                }

                if ($cell === self::SPACE_CELL) {
                    $flushCurrentCells();
                    $pushWordToken();
                    $tokens[] = ['kind' => 'space'];
                    continue;
                }

                $currentCells[] = $cell;
            }

            $flushCurrentCells();
        };

        $appendAtomicCells = function (array $cells) use (&$tokens, $pushSegment, $pushWordToken): void {
            $currentCells = [];

            $flushCurrentCells = function () use (&$currentCells, $pushSegment): void {
                $pushSegment(true, $currentCells);
                $currentCells = [];
            };

            foreach ($cells as $cell) {
                if ($cell === "\n") {
                    $flushCurrentCells();
                    $pushWordToken();
                    $tokens[] = ['kind' => 'newline'];
                    continue;
                }

                $currentCells[] = $cell;
            }

            $flushCurrentCells();
        };

        foreach ($parts as $part) {
            $cells = self::tokenizeDisplayCells($part['template']);
            if ($part['wrap'] === 'never') {
                $appendAtomicCells($cells);
                continue;
            }

            $appendNormalCells($cells);
        }

        $pushWordToken();

        return $tokens;
    }

    private static function measureWordWidth(array $token): int
    {
        return array_reduce(
            $token['segments'],
            fn(int $width, array $segment) => $width + count($segment['cells']),
            0
        );
    }

    private static function wordTokenToString(array $token): string
    {
        return implode('', array_map(
            fn(array $segment) => implode('', $segment['cells']),
            $token['segments']
        ));
    }

    private static function takeWordPrefix(array $token, int $maxWidth): array
    {
        if ($maxWidth <= 0) {
            return [
                'prefix' => null,
                'remainder' => self::createWordToken(
                    self::cloneSegments($token['segments'])
                ),
            ];
        }

        $prefixSegments = [];
        $remainderSegments = [];
        $consumedWidth = 0;
        $segments = $token['segments'];

        $appendRemainder = function (int $startIndex) use (&$remainderSegments, $segments): void {
            $remainderSegments = array_merge(
                $remainderSegments,
                self::cloneSegments(array_slice($segments, $startIndex))
            );
        };

        foreach ($segments as $index => $segment) {
            $segmentWidth = count($segment['cells']);

            if ($consumedWidth + $segmentWidth <= $maxWidth) {
                $prefixSegments[] = self::cloneSegment($segment);
                $consumedWidth += $segmentWidth;
                continue;
            }

            if ($segment['atomic']) {
                if ($consumedWidth === 0) {
                    $prefixSegments[] = [
                        'atomic' => true,
                        'cells' => array_slice($segment['cells'], 0, $maxWidth),
                    ];

                    if ($segmentWidth > $maxWidth) {
                        $remainderSegments[] = [
                            'atomic' => true,
                            'cells' => array_slice($segment['cells'], $maxWidth),
                        ];
                    } else {
                        $remainderSegments[] = self::cloneSegment($segment);
                    }

                    $appendRemainder($index + 1);
                } else {
                    $remainderSegments[] = self::cloneSegment($segment);
                    $appendRemainder($index + 1);
                }

                return [
                    'prefix' => !empty($prefixSegments)
                        ? self::createWordToken($prefixSegments)
                        : null,
                    'remainder' => !empty($remainderSegments)
                        ? self::createWordToken($remainderSegments)
                        : null,
                ];
            }

            $remainingWidth = $maxWidth - $consumedWidth;
            if ($remainingWidth > 0) {
                $prefixSegments[] = [
                    'atomic' => false,
                    'cells' => array_slice($segment['cells'], 0, $remainingWidth),
                ];
                $remainderSegments[] = [
                    'atomic' => false,
                    'cells' => array_slice($segment['cells'], $remainingWidth),
                ];
            } else {
                $remainderSegments[] = self::cloneSegment($segment);
            }

            $appendRemainder($index + 1);

            return [
                'prefix' => !empty($prefixSegments)
                    ? self::createWordToken($prefixSegments)
                    : null,
                'remainder' => !empty($remainderSegments)
                    ? self::createWordToken($remainderSegments)
                    : null,
            ];
        }

        return [
            'prefix' => self::createWordToken($prefixSegments),
            'remainder' => null,
        ];
    }

    private static function chunkWordTokens(int $width, array $tokens): array
    {
        $chunkedTokens = [];

        foreach ($tokens as $token) {
            if ($token['kind'] !== 'word' || self::measureWordWidth($token) <= $width) {
                $chunkedTokens[] = $token;
                continue;
            }

            $remaining = $token;
            while ($remaining !== null && self::measureWordWidth($remaining) > $width) {
                $split = self::takeWordPrefix($remaining, $width);
                if ($split['prefix'] === null) {
                    break;
                }

                $chunkedTokens[] = $split['prefix'];
                $remaining = $split['remainder'];
            }

            if ($remaining !== null) {
                $chunkedTokens[] = $remaining;
            }
        }

        return $chunkedTokens;
    }

    private static function wrapTemplateTokens(int $width, array $tokens): array
    {
        $lines = [[
            'line' => '',
            'length' => 0,
        ]];
        $chunkedTokens = self::chunkWordTokens($width, $tokens);

        foreach ($chunkedTokens as $index => $token) {
            $lastIndex = count($lines) - 1;
            $lineLength = $lines[$lastIndex]['length'];
            $emptyLine = ($lines[$lastIndex]['line'] === '');

            if (
                $token['kind'] === 'newline'
                && $index > 0
                && $chunkedTokens[$index - 1]['kind'] === 'newline'
            ) {
                $lines[] = [
                    'line' => '',
                    'length' => 0,
                ];
                continue;
            }

            if ($token['kind'] === 'newline') {
                $lines[$lastIndex] = [
                    'line' => $lines[$lastIndex]['line'],
                    'length' => $width,
                ];
                continue;
            }

            if ($token['kind'] === 'space') {
                if ($lineLength + 1 > $width) {
                    continue;
                }

                $lines[$lastIndex] = [
                    'line' => $lines[$lastIndex]['line'] . self::SPACE_CELL,
                    'length' => $lineLength + 1,
                ];
                continue;
            }

            $wordWidth = self::measureWordWidth($token);
            if ($width >= $wordWidth + $lineLength && !$emptyLine) {
                $lines[$lastIndex] = [
                    'line' => $lines[$lastIndex]['line'] . self::wordTokenToString($token),
                    'length' => $lineLength + $wordWidth,
                ];
                continue;
            }

            $lines[] = [
                'line' => self::wordTokenToString($token),
                'length' => $wordWidth,
            ];
        }

        if ($lines[0]['line'] === '') {
            $lines = array_slice($lines, 1);
        }

        return array_map(fn(array $line) => $line['line'], $lines);
    }
}
